datafusion-examples/examples/sql_analysis.rs

```diff
@@ -0,0 +1,152 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This example shows how we can use the structures that DataFusion provide to perform
+//! SQL Analysis.
+//! We'll show how we can count the amount of join's in a query as well as how many
+//! join tree's there are
+
+use std::{fs, sync::Arc};
+
+use datafusion::common::Result;
+use datafusion::{
+    datasource::MemTable,
+    execution::context::{SessionConfig, SessionContext},
+};
+use datafusion_common::tree_node::{TreeNode, TreeNodeRecursion};
+use datafusion_expr::LogicalPlan;
+use test_utils::tpcds::tpcds_schemas;
+
+/// Counts the total amount of joins in a plan
+fn total_join_count(plan: &LogicalPlan) -> usize {
+    let mut total = 0;
+
+    // We can use the TreeNode API to walk over a LogicalPlan.
+    plan.apply(|node| {
+        // if we encouter a join we update the total
+        if matches!(node, LogicalPlan::Join(_) | LogicalPlan::CrossJoin(_)) {
+            total += 1;
+        }
+        Ok(TreeNodeRecursion::Continue)
+    })
+    .unwrap();
+
+    total
+}
+
+/// Counts the total amount of joins and collects every join tree with it's join count
+/// The list of groups sums to the total count
+fn count_trees(plan: &LogicalPlan) -> (usize, Vec<usize>) {
```

Comment:
I found it a little hard to follow what qualified as a join tree and when they are different

maybe a definition and few text examples would help, like

```
A join tree is defined as the largest sub trees which consistt entirely of joins

For example

  JOIN
  /  \
A   JOIN
     /  \
    B    C

has a single join tree (A-B-C) but this plan:

  JOIN
  /  \
A   GROUP
       |
      JOIN
     /  \
    B    C

Has two join trees (A, B-C)
```